pandas/core/dtypes/cast.py

```diff
@@ -1314,6 +1315,37 @@ def find_result_type(left_dtype: DtypeObj, right: Any) -> DtypeObj:
         #  which will make us upcast too far.
         if lib.is_float(right) and right.is_integer() and left_dtype.kind != "f":
             right = int(right)
+        # After NEP 50, numpy won't inspect Python scalars
+        # TODO: do we need to recreate numpy's inspection logic for floats too
+        # (this breaks some tests)
+        if isinstance(right, int) and not isinstance(right, np.integer):
+            # This gives an unsigned type by default
+            # (if our number is positive)
+
+            # If our left dtype is signed, we might not want this since
+            # this might give us 1 dtype too big, though.
+            # We should check if the corresponding int dtype (e.g. int64 for uint64)
+            # can hold the number by a little hack where we ask numpy for the min
+            # type of the negative of the number (which can't return an unsigned)
+            # If the dtype is the same size, then we will use that
+            right_dtype = np.min_scalar_type(right)
+            if right == 0:
+                # Special case 0, our trick will not work for it since
+                # np.min_scalar_type(-0) will still give something unsigned
+                right = left_dtype
+            elif right > 0 and not np.issubdtype(left_dtype, np.unsignedinteger):
```

Comment:
I think this would be simpler if you just did:

```python
elif right > 0 and right <= 2 ** (8 * left_dtype.itemsize - 1) - 1:
    # use signed dtype
```

would cut down on some of the branching

Reply:
Hm, do you know a way to programatically convert from the unsigned dtype to the signed one?

I think that was why I went with my hack in the first place.

Reply:
I think you could create the new type by doing something like `f"i{left_dtype.itemsize}"`

Reply:
Nice! Updated the code.